### PR TITLE
KEYCLOAK-12243 MultivaluedHashMap.equalsIgnoreValueOrder() returns true for some maps that are not equal

### PR DESCRIPTION
**Pull request by tkyjovsk**
fix + test

### CODE REVIEW

**Comment by toziol** [comment]
You should add this test :

        map.clear();
        map.put(1, Arrays.asList(new Integer[]{1, 2, 3}));
        differentMap.clear();
        differentMap.put(1, Arrays.asList(new Integer[]{1, 2, 3}));
        differentMap.put(2, Arrays.asList(new Integer[]{4, 5, 6}));
        equalsIgnoreValueOrder_shouldReturnFalseForDifferentMaps(map, differentMap); // diff entrySet size

File: common/src/main/java/org/keycloak/common/util/MultivaluedHashMap.java [hunk 1]
```diff
@@ -101,11 +101,11 @@ public V getFirst(K key)
    {
       List<V> list = get(key);
       if (list == null)
-         put(key, list = new ArrayList<V>());
+         put(key, list = new ArrayList<>());
       return list;
    }
 
-   public void addAll(MultivaluedHashMap<K, V> other)
+   public final void addAll(MultivaluedHashMap<K, V> other)
```
Comment (hmlnarik) [inline common/src/main/java/org/keycloak/common/util/MultivaluedHashMap.java]: Why `final`?

Comment (tkyjovsk) [reply common/src/main/java/org/keycloak/common/util/MultivaluedHashMap.java]: Method used in the constructor shouldn't be overridable.

File: common/src/test/java/org/keycloak/common/util/MultivaluedHashMapTest.java [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        MultivaluedHashMap equalMap = new MultivaluedHashMap();
+        map.put(1, Arrays.asList(new Integer[]{1, 2, 3}));
+        equalMap.put(1, Arrays.asList(new Integer[]{3, 2, 1}));
+        equalsIgnoreValueOrder_shouldReturnTrueForEqualMaps(map, equalMap);
+
+        map.put(2, Arrays.asList(new Integer[]{4, 5, 6}));
+        equalMap.put(2, Arrays.asList(new Integer[]{5, 6, 4}));
+        equalsIgnoreValueOrder_shouldReturnTrueForEqualMaps(map, equalMap);
+
+        map.put("c", Arrays.asList(new String[]{"x", "y", "z"}));
```
Comment (fperot74) [inline common/src/test/java/org/keycloak/common/util/MultivaluedHashMapTest.java]: This is not false but I'm mitigated about this test : map was first initialized like if it was a MultivaluedHashMap<Integer, Integer>...
Here it is used like if it was a MultivaluedHashMap<String, String>. Mixing types is not what the object has been created for but it is not considered as an error...
Ideally, MultivaluedHashMap should not have mixed type content and types should be provided at declaration... but here, it is just for test purpose.

Comment (tkyjovsk) [reply common/src/test/java/org/keycloak/common/util/MultivaluedHashMapTest.java]: Yes, it was just for the test. But you're right, the types shouldn't be mixed. I addressed your suggestions in the next push. Thanks!

File: common/src/test/java/org/keycloak/common/util/MultivaluedHashMapTest.java [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        map.put(2, Arrays.asList(new Integer[]{4, 5, 6, 7, 8}));
+        equalMap.clear();
+        equalMap.put(1, Arrays.asList(new Integer[]{4, 3, 2, 6, 1}));
+        equalMap.put(2, Arrays.asList(new Integer[]{6, 7, 4, 8, 5}));
+        equalsIgnoreValueOrder_shouldReturnFalseForDifferentMaps(map, equalMap);
+    }
+
+    @Test
+    public void testEqualsIgnoreValueOrder_sameLengthDifferentValues() {
+        MultivaluedHashMap map = new MultivaluedHashMap();
```
[the remark this answers is not in the document]
Comment (tkyjovsk) [reply common/src/test/java/org/keycloak/common/util/MultivaluedHashMapTest.java]: Resolved in `4e01018`.

File: common/src/test/java/org/keycloak/common/util/MultivaluedHashMapTest.java [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+                map.equalsIgnoreValueOrder(differentMap));
+        assertFalse(String.format("MultivaluedHashMap.equalsIgnoreValueOrder() should return `false` for maps that are different. \nmap1: %s \nmap2: %s", differentMap, map),
+                differentMap.equalsIgnoreValueOrder(map));
+    }
+
+    @Test
+    public void testEqualsIgnoreValueOrder_exactlyEqual() {
+        MultivaluedHashMap map = new MultivaluedHashMap();
+        MultivaluedHashMap equalMap = new MultivaluedHashMap();
+        map.put(1, Arrays.asList(new Integer[]{1, 2, 3}));
```
[the remark this answers is not in the document]
Comment (tkyjovsk) [reply common/src/test/java/org/keycloak/common/util/MultivaluedHashMapTest.java]: Thanks for this suggestion. Resolved in `4e01018`.

File: common/src/test/java/org/keycloak/common/util/MultivaluedHashMapTest.java [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        equalMap.put(3, Arrays.asList(9, 7, 8));
+        equalsIgnoreValueOrder_shouldReturnTrueForEqualMaps(map, equalMap);
+
+        map.clear();
+        equalMap.clear();
+        map.put(1, Arrays.asList(1, 2, 3, 4, 5));
+        map.put(2, Arrays.asList(4, 5, 6, 7, 8));
+        equalMap.put(1, Arrays.asList(4, 3, 2, 6, 1));
+        equalMap.put(2, Arrays.asList(6, 7, 4, 8, 5));
+        equalsIgnoreValueOrder_shouldReturnFalseForDifferentMaps(map, equalMap);
```
Comment (tkyjovsk) [inline common/src/test/java/org/keycloak/common/util/MultivaluedHashMapTest.java]: This needs to be fixed to:
```
        equalMap.put(1, Arrays.asList(4, 3, 2, 5, 1));
        equalMap.put(2, Arrays.asList(6, 7, 4, 8, 5));
        equalsIgnoreValueOrder_shouldReturnTrueForEqualMaps(map, equalMap);
```

File: common/src/main/java/org/keycloak/common/util/MultivaluedHashMap.java (hunk 1 above)
Comment (hmlnarik) [reply common/src/main/java/org/keycloak/common/util/MultivaluedHashMap.java]: Thanks for explanation. I did not look around the PR.

File: common/src/test/java/org/keycloak/common/util/MultivaluedHashMapTest.java [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+/**
+ *
+ * @author tkyjovsk
+ */
+public class MultivaluedHashMapTest {
+
+    public void equalsIgnoreValueOrder_shouldReturnTrueForEqualMaps(MultivaluedHashMap map, MultivaluedHashMap equalMap) {
```
Comment (fperot74) [inline common/src/test/java/org/keycloak/common/util/MultivaluedHashMapTest.java]: You can add <T, R> in both equalsIgnoreValueOrder_XXX methods like this:
private <T, R> void equalsIgnoreValueOrder_shouldReturnTrueForEqualMaps(MultivaluedHashMap<T, R> map, MultivaluedHashMap<T, R> equalMap) {

**Review by fperot74 — DISMISSED** [review]

File: common/src/test/java/org/keycloak/common/util/MultivaluedHashMapTest.java (hunk 6 above)
Comment (tkyjovsk) [reply common/src/test/java/org/keycloak/common/util/MultivaluedHashMapTest.java]: I added the generics, thanks.

File: common/src/main/java/org/keycloak/common/util/MultivaluedHashMap.java [hunk 7]
```diff
@@ -47,7 +47,7 @@ public MultivaluedHashMap(MultivaluedHashMap<K, V> config) {
 
    public void putSingle(K key, V value)
    {
-      List<V> list = new ArrayList<V>();
+      List<V> list = new ArrayList<>();
```
Comment (mposolda) [inline common/src/main/java/org/keycloak/common/util/MultivaluedHashMap.java]: This is in the "common" module, which is used by adapters etc and require to compile on some historic java versions (JDK 7 etc). Due this, I would not use `ArrayList<>` , but rather keep this line as  `ArrayList<V>`  . As it was before

File: common/src/main/java/org/keycloak/common/util/MultivaluedHashMap.java [hunk 8]
```diff
@@ -101,11 +101,11 @@ public V getFirst(K key)
    {
       List<V> list = get(key);
       if (list == null)
-         put(key, list = new ArrayList<V>());
+         put(key, list = new ArrayList<>());
```
Comment (mposolda) [inline common/src/main/java/org/keycloak/common/util/MultivaluedHashMap.java]: Same as above

File: common/src/test/java/org/keycloak/common/util/MultivaluedHashMapTest.java [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+        map.put(1, Arrays.asList(1, 2, 3));
+        map.put(2, Arrays.asList(4, 5, 6));
+        differentMap.put(1, Arrays.asList(1, 2, 3));
+        differentMap.put(2, Arrays.asList(4, 5, 5, 4, 6));
+        equalsIgnoreValueOrder_shouldReturnFalseForDifferentMaps(map, differentMap);
+    }
+
+    @Test
+    public void testEqualsIgnoreValueOrder_differentLengthDifferentValues() {
+        MultivaluedHashMap<Integer, Integer> map = new MultivaluedHashMap<>();
```
Comment (mposolda) [inline common/src/test/java/org/keycloak/common/util/MultivaluedHashMapTest.java]: Same as above

File: common/src/test/java/org/keycloak/common/util/MultivaluedHashMapTest.java [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+    public <T, R> void equalsIgnoreValueOrder_shouldReturnFalseForDifferentMaps(MultivaluedHashMap<T, R> map, MultivaluedHashMap<T, R> differentMap) {
+        assertFalse(String.format("MultivaluedHashMap.equalsIgnoreValueOrder() should return `false` for maps that are different. \nmap1: %s \nmap2: %s", map, differentMap),
+                map.equalsIgnoreValueOrder(differentMap));
+        assertFalse(String.format("MultivaluedHashMap.equalsIgnoreValueOrder() should return `false` for maps that are different. \nmap1: %s \nmap2: %s", differentMap, map),
+                differentMap.equalsIgnoreValueOrder(map));
+    }
+
+    @Test
+    public void testEqualsIgnoreValueOrder_exactlyEqual() {
+        MultivaluedHashMap<Integer, Integer> map = new MultivaluedHashMap<>();
```
Comment (mposolda) [inline common/src/test/java/org/keycloak/common/util/MultivaluedHashMapTest.java]: Same as above

**Review by mposolda — CHANGES_REQUESTED** [review]
I've added few minor comments, but all of them are for the same thing. If it is possible, I would avoid using construct like `List<V> l = new ArrayList<>();` or `Map<K,V> m = new HashMap<>()` , but keep it everywhere in the old format to ensure it can safely compile on historic java versions. That applies to test as well.

Besides that, it looks good and I would approve once @fperot74 approves.

File: common/src/main/java/org/keycloak/common/util/MultivaluedHashMap.java (hunk 7 above)
Comment (tkyjovsk) [reply common/src/main/java/org/keycloak/common/util/MultivaluedHashMap.java]: Java 7 should support type inference for generics. I tried compiling the module with JDK 7 and it compiled fine. (There was some problem with surefire plugin but that seems unrelated.)

Comment (mposolda) [reply common/src/main/java/org/keycloak/common/util/MultivaluedHashMap.java]: Cool, Thanks! I don't know why I though it doesn't :) Approving and merging then.

**Review by mposolda — APPROVED** [review]

**Pull request merged**